src/action-emitter.js

```diff
@@ -19,24 +20,16 @@ export function makeActionEmitter(onPotentiallyUnhandledAction) {
             });
         },
         emit(action) {
-            const listenersToFire = [];
-            for (let i = 0; i < listeners.length; i++) {
-                const listener = listeners[i];
-                const isValid = listener.patternChecker(listener.pattern, listener.pattern2ndArg, action);
+            const matchingListeners = listeners.filter((listener) => listener.patternChecker(listener.pattern, listener.pattern2ndArg, action));
 
-                // remove the listener first to avoid the callback firing
-                // an action that would try and restart the saga
-                if (isValid) {
-                    listenersToFire.push(listener);
-                    listeners.splice(i, 1);
-                    i--;
-                }
-            }
-
-            if (listenersToFire.every(({ pattern }) => pattern === '*' || pattern.isLoose)) {
+            if (matchingListeners.every(({ pattern }) => pattern === '*' || pattern.isLoose)) {
                 onPotentiallyUnhandledAction(action);
             }
 
+            // avoid the callback firing an action that would try and restart the saga
+            const listenersToFire = matchingListeners.filter((listener) => !suspendedListeners.includes(listener));
+            suspendedListeners.push(...listenersToFire);

```

Comment:
what clears suspended listeners?

its not really "suspended" listeners, its just an approach to stop a listener being fired twice

Reply:
It's rebuilt at some point - in the existing code, nothing restored the removed listener back into `listeners` either.

Yes I saw the purpose, I named it based on "we are firing this now so it's suspended from further firing". Maybe `activeListeners` or `firingListeners` ?